JASP-Desktop/utilities/languagemodel.cpp

```diff
@@ -157,59 +159,31 @@ void LanguageModel::resultsPageLoaded()
 
 QString LanguageModel::getLocalName(QLocale::Language cl) const
 {
-	LanguageInfo li = _languagesInfo[cl];
-	return li.localName;
-
+	return _languagesInfo[cl].localName;
 }
 
 QString LanguageModel::getNativeLanguaName(QLocale::Language cl) const
 {
-	LanguageInfo li = _languagesInfo[cl];
-	return li.nativeLanguageName;
-
-}
-
-QLocale::Language LanguageModel::getLanguageKeyFromName(QString languagename) const
-{
-	QMapIterator<QLocale::Language, LanguageInfo> it(_languagesInfo) ;
-
-	while (it.hasNext())
-	{
-		it.next();
-		QLocale::Language l =  it.key();
-		LanguageInfo li =_languagesInfo[l];
-		if (languagename == li.nativeLanguageName) return l;
-	}
-	return QLocale::English; //By default
-
+	return _languagesInfo[cl].nativeLanguageName;
 }
 
 QString  LanguageModel::getLocalNameFromQmFileName(QString filename) const
 {
+	int		start		= filename.indexOf('_');
+	QString localname	= filename.mid(start + 1);
+	int		end			= localname.lastIndexOf('.');
 
-	int start = filename.indexOf('_');
-	QString localname = filename.mid(start + 1);
-
-	int end = localname.lastIndexOf('.');
-	localname = localname.left(end);
-
-	return localname;
+	return localname.left(end);
 
 }
 
 bool LanguageModel::isValidLocalName(QString filename, QLocale::Language & cl)
 {
+	QLocale loc;
 
-	//Checks if the filename has a proper localename suffix for a valid CLocale::Language
-	//https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes
-	//We use the two lettecode in the qm filename
-
-	QString localname = getLocalNameFromQmFileName(filename);
-	QLocale loc(localname);
-	QString languagename = "";
-	languagename = loc.nativeLanguageName();
-	if (languagename == "")
+	if(!isValidLocalName(filename, loc))
```

Comment:
According to my c++ knowledge is the method
`bool LanguageModel::isValidLocalName(QString filename, QLocale & local)
{
	//Checks if the filename has a proper localename suffix for a valid CLocale::Language
	//https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes
	//We use the two lettecode in the qm filename

	QLocale loc			 = getLocalNameFromQmFileName(filename);
	QString languagename = loc.nativeLanguageName();

	if (languagename == "")
		return false;

	local = loc;
	return true;

}`

really wrong: local get a local value. I don't know how it can work

Reply:
I do not understand what you are trying to say here, and whatever you copied there is a bit unclear?

Maybe it helps to look at [the entire file](https://github.com/jasp-stats/jasp-desktop/blob/bd35adce2dd8c6a184a16022fd90459ed39c8e7a/JASP-Desktop/utilities/languagemodel.cpp)?

Then we see:
```C++
bool LanguageModel::isValidLocalName(QString filename, QLocale::Language & cl)
{
	QLocale loc;

	if(!isValidLocalName(filename, loc))
		return false;

	cl = loc.language();
	return true;

}

bool LanguageModel::isValidLocalName(QString filename, QLocale & local)
{
	//Checks if the filename has a proper localename suffix for a valid CLocale::Language
	//https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes
	//We use the two lettecode in the qm filename

	QLocale loc			 = getLocalNameFromQmFileName(filename);
	QString languagename = loc.nativeLanguageName();

	if (languagename == "")
		return false;

	local = loc;
	return true;

}
```

Reply:
First the code (and comment) were duplicated, which I thought a bit ugly. And then I noticed that it can be reused.

If you then look at `local = loc;` then that means that local gets a copy of loc right?
Because it is already a reference to whatever was outside the function?

Reply:
See also https://isocpp.org/wiki/faq/references

How could you ever assign anything to a reference if it would change the address the reference points to?

Reply:
Sorry, it is something I would never do (local = loc) with an object reference, I thought 'local' would take the reference of 'loc' local variable. But it is a copy indeed. Usually if I want to copy a local variable, I return it, but it can be done this way also.